### PR TITLE
[ROCm] re-enable support RCCL with fixes

### PR DESCRIPTION
**Pull request by i-chaochen**
This one is based on https://github.com/openxla/xla/pull/8689 with 

1. collective bugs fix based on https://github.com/openxla/xla/commit/975a85a697cc206cf969cb5e0c24d3d06701ab81#diff-e791b792baae6f61e2050f5ef3b44f8c9c3e88accd980cf70d350e5ba99a5474R589-R595
2. ~~compile without speficic platform https://github.com/openxla/xla/pull/8689/files#r1463020991~~

### CODE REVIEW

**Comment by jglaser** [comment]
Thanks for picking up slack :)

**Comment by i-chaochen** [comment]
@xla-rotation please check whether this is ok now. (most parts are approved in https://github.com/openxla/xla/pull/8689)

**Review by akuegel — APPROVED** [review]

**Review by sgerrard — APPROVED** [review]

**Comment by sgerrard** [comment]
Please add a "TESTED=" field to the PR description in order for the submit to complete.

**Review by ddunl — APPROVED** [review]

File: xla/stream_executor/build_defs.bzl
Comment (akuegel) [inline xla/stream_executor/build_defs.bzl:23]: I think we still need that macro, it is used for example for some targets in the tools directory.

**Review by akuegel — CHANGES_REQUESTED** [review]
Sorry, I missed that you were deleting the if_cuda_or_rocm macro. We still need that one (although ideally we shouldn't).

File: xla/stream_executor/build_defs.bzl
Comment (i-chaochen) [reply xla/stream_executor/build_defs.bzl]: Thanks for notice! Added it back.

**Review by akuegel — APPROVED** [review]

**Comment by i-chaochen** [comment]
hmmm....XLA Linux ARM64 CPU is failed now....shouldn't be related to this bazel file.

```
2024-02-06 11:43:23.787069: F xla/service/rendezvous.cc:105] Check failed: state != kPending rendezvous can't remain in pending state after state update failure
*** Received signal 6 ***
*** BEGIN MANGLED STACK TRACE ***
2024-02-06 11:43:23.787191: F xla/service/rendezvous.cc:105] Check failed: state != kPending rendezvous can't remain in pending state after state update failure
2024-02-06 11:43:23.787246: F xla/service/rendezvous.cc:105] Check failed: state != kPending rendezvous can't remain in pending state after state update failure
2024-02-06 11:43:23.787296: F xla/service/rendezvous.cc:105] Check failed: state != kPending rendezvous can't remain in pending state after state update failure
2024-02-06 11:43:23.787362: F xla/service/rendezvous.cc:105] Check failed: state != kPending rendezvous can't remain in pending state after state update failure
================================================================================

```

File: xla/pjrt/gpu/BUILD
Comment (akuegel) [inline xla/pjrt/gpu/BUILD:5]: Maybe also revert changes to this file. It seems that the if_cuda_or_rocm macro from stream_executor:build_defs.bzl is the same as if_gpu_is_configured (surprisingly). So changing this macro load is a behavior change.

**Review by akuegel — APPROVED** [review]
Thanks :)

**Comment by i-chaochen** [comment]
@akuegel the feedback/copybara is still red cross, but I think it should be ok now? (fingers crossed)

**Review by sgerrard — APPROVED** [review]

File: xla/service/gpu/BUILD
Comment (ezhulenev) [inline xla/service/gpu/BUILD:958]: We need to keep this indirection via stub because NCCL is not always enabled and everything breaks without it

File: xla/service/gpu/nccl_api.cc
Comment (ezhulenev) [inline xla/service/gpu/nccl_api.cc:546]: I'm trying to delete `XLA_ENABLE_XCCL` ifdef from XLA code base, and new coding style is "no new ifdefs", the plan to is keep "real" implementation completely separate from a stub, albeit it adds some extra complexity to build files.

We certainly don't want large blocks of code undef ifdef as it breaks all our static analysis tooling. It's ok to have small regions under ifdef to define aliases, but not in the middle of the file.

Comment (akuegel) [reply xla/service/gpu/nccl_api.cc:546]: But that means we will need a if_gpu_is_configured macro that supports an else branch.

Comment (akuegel) [reply xla/service/gpu/nccl_api.cc:546]: Right now, if_gpu_is_configured is if_cuda_is_configured(x) + if_rocm_is_configured(x). For if_cuda_or_rocm, we simply specify a select that is the union of if_cuda and if_rocm macro. For if_gpu_is_configured, that is not easily possible, as if_cuda_is_configured and if_rocm_is_configured are in template .bzl files and do something differently based on build variables: "if %{rocm_is_configured}" "if %{cuda_is_configured}". I think we would need to adapt cuda_configure.bzl and rocm_configure.bzl to specify both variables, so that we can make use of them.

Comment (akuegel) [reply xla/service/gpu/nccl_api.cc:546]: And thinking more about it, I don't know how this can work. Maybe someone can figure this out, but it will be very tricky.

Comment (akuegel) [reply xla/service/gpu/nccl_api.cc:546]: I am trying https://github.com/openxla/xla/pull/9266 now, which loads enable_cuda from cuda_configure.bzl and then sets a if_gpu_is_configured macro to false only if also enable_cuda returns false. So we can have the if_gpu_is_configured macro in the rocm:build_defs.bzl file, and re-export if via the stream_executor build_defs.bzl file.

Comment (akuegel) [reply xla/service/gpu/nccl_api.cc:546]: The PR has been merged now, can you please try to switch to if_gpu_is_configured macro and keep the stub file?

Comment (i-chaochen) [reply xla/service/gpu/nccl_api.cc]: Thanks for the help! Done.

**Comment by i-chaochen** [comment]
hmmm....don't know why elemental_ir_emitter_test_gpu is failed here, I tried it in my A100 and it's no problem at all.

```
[ RUN      ] ElementalIrEmitterExecutionTestWithoutFastMinMax.MinimumHandlesNaNsOnTheRight
2024-02-08 12:15:47.591851: I xla/tests/literal_test_util.cc:56] expected: f32[5,5] f32[5,5] {
  { -nan, -nan, -nan, -nan, -nan },
  { -nan, -nan, -nan, -nan, -nan },
  { -nan, -nan, -nan, -nan, -nan },
  { -nan, -nan, -nan, -nan, -nan },
  { -nan, -nan, -nan, -nan, -nan }
}
2024-02-08 12:15:47.591899: I xla/tests/literal_test_util.cc:58] actual:   f32[5,5] f32[5,5] {
  { -1, -1, -1, -1, -1 },
  { -1, -1, -1, -1, -1 },
  { -1, -1, -1, -1, -1 },
  { -1, -1, -1, -1, -1 },
  { -1, -1, -1, -1, -1 }
}
```

it doesn't seem related to what I have changed...should I rebase and rerun it again?

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
> hmmm....don't know why elemental_ir_emitter_test_gpu is failed here, I tried it in my A100 and it's PASSED.
> 
> ```
> [ RUN      ] ElementalIrEmitterExecutionTestWithoutFastMinMax.MinimumHandlesNaNsOnTheRight
> 2024-02-08 12:15:47.591851: I xla/tests/literal_test_util.cc:56] expected: f32[5,5] f32[5,5] {
>   { -nan, -nan, -nan, -nan, -nan },
>   { -nan, -nan, -nan, -nan, -nan },
>   { -nan, -nan, -nan, -nan, -nan },
>   { -nan, -nan, -nan, -nan, -nan },
>   { -nan, -nan, -nan, -nan, -nan }
> }
> 2024-02-08 12:15:47.591899: I xla/tests/literal_test_util.cc:58] actual:   f32[5,5] f32[5,5] {
>   { -1, -1, -1, -1, -1 },
>   { -1, -1, -1, -1, -1 },
>   { -1, -1, -1, -1, -1 },
>   { -1, -1, -1, -1, -1 },
>   { -1, -1, -1, -1, -1 }
> }
> ```
> 
> it doesn't seem related to what I have changed...should I rebase and rerun it again?

That failure is unrelated, it got broken earlier today.

**Comment by i-chaochen** [comment]
Hi @ezhulenev I have changed stub back, wondering is this PR ok now? Thanks!

File: xla/service/gpu/nccl_api_stub.cc
Comment (akuegel) [inline xla/service/gpu/nccl_api_stub.cc:2]: I guess removing the blank lines here is not intended?

**Comment by akuegel** [comment]
> Hi @ezhulenev I have changed stub back, wondering is this PR ok now? Thanks!

This got blocked on missing internal approval. But I also noticed one minor thing regarding the stub file and added a comment.

File: xla/service/gpu/nccl_api_stub.cc
Comment (i-chaochen) [reply xla/service/gpu/nccl_api_stub.cc]: Sure :)  Done

File: xla/service/gpu/nccl_api_stub.cc
Comment (i-chaochen) [reply xla/service/gpu/nccl_api_stub.cc:6]: Thanks for notice. I directly copied and pasted this file and it's sloppy.

**Review by akuegel — APPROVED** [review]

**Comment by i-chaochen** [comment]
I have rebased this PR due to runtime refactoring.

**Review by akuegel — APPROVED** [review]

**Comment by i-chaochen** [comment]
Hi @ddunl this is one has been stuck for a long while, wonder is it ok to be merged now?

**Comment by i-chaochen** [comment]
gentle ping @xla-rotation

**Pull request closed without merging**
